File: 第3合集-AI时代编程新范式：Vibe Coding/260306-大模型编程能力实测/02-模型源码/GLM-5/backend/api/danmaku.py
"""
弹幕相关API路由
"""
from fastapi import APIRouter, HTTPException, Query
from fastapi.responses import JSONResponse
from pydantic import BaseModel
from typing import Optional

from services.danmaku_service import danmaku_service
from models.schemas import APIResponse

router = APIRouter(prefix="/danmaku", tags=["弹幕"])


class SendDanmakuRequest(BaseModel):
    """发送弹幕请求"""
    oid: int
    message: str
    time: float
    color: int = 16777215  # 默认白色
    fontsize: int = 25
    mode: int = 1  # 1为滚动弹幕


@router.get("/list")
async def get_danmaku(
    cid: int = Query(..., description="视频CID"),
    bvid: Optional[str] = Query(None, description="视频BV号"),
    aid: Optional[int] = Query(None, description="视频AV号")
):
    """获取弹幕列表"""
    if not bvid and not aid:
        raise HTTPException(status_code=400, detail="请提供bvid或aid")

    danmaku_list = await danmaku_service.get_danmaku(cid=cid, bvid=bvid, aid=aid)
    return APIResponse(success=True, data=[d.model_dump() for d in danmaku_list])


@router.post("/send")
async def send_danmaku(request: SendDanmakuRequest):
    """发送弹幕"""
    result = await danmaku_service.send_danmaku(
        oid=request.oid,
        message=request.message,
        time=request.time,
        color=request.color,
        fontsize=request.fontsize,
        mode=request.mode
    )
    if result.get('success'):
        return APIResponse(success=True, message=result.get('message', '发送成功'))
    return APIResponse(success=False, message=result.get('message', '发送失败'))

File: 第3合集-AI时代编程新范式：Vibe Coding/260306-大模型编程能力实测/02-模型源码/Step-3.5-Flash/bilibili-clone/backend/src/routes/danmaku.ts
import { Router } from 'express';
import { bilibiliApi } from '../services/bilibili-api';
import { requireAuth } from '../middleware/auth';

const router = Router();

router.get('/video/:bvid', async (req, res) => {
  try {
    const { bvid } = req.params;
    const { cid } = req.query;

    if (!cid) {
      return res.status(400).json({
        code: -1,
        message: '缺少cid参数',
        data: null
      });
    }

    const danmakuList = await bilibiliApi.getDanmaku(bvid, parseInt(cid as string));
    res.json({
      code: 0,
      message: 'success',
      data: danmakuList
    });
  } catch (error: any) {
    res.status(500).json({
      code: -1,
      message: error.message || '获取弹幕失败',
      data: null
    });
  }
});

router.post('/video/:bvid', requireAuth, async (req, res) => {
  try {
    const { bvid } = req.params;
    const { cid, content, time, color } = req.body;

    if (!cid || !content || time === undefined) {
      return res.status(400).json({
        code: -1,
        message: '参数不完整',
        data: null
      });
    }

    const result = await bilibiliApi.sendDanmaku(bvid, parseInt(cid), content, time, color);
    res.json({
      code: result.code,
      message: result.message,
      data: result.data
    });
  } catch (error: any) {
    res.status(500).json({
      code: -1,
      message: error.message || '发送弹幕失败',
      data: null
    });
  }
});

export default router;